File: framework-ts/module/devian-domain-game/features/ability/AbilityUnitHero.ts
import { EQUIP_SLOT_TYPE, UNIT_STAT_TYPE, UNIT_HERO, UNIT_HERO_LEVEL } from '../../Generated/Game.g';
import { AbilityItemEquip } from './AbilityItemEquip';
import { AbilityEquipPlacementFailure, AbilityEquipSlotPolicy } from './AbilityEquipSlotPolicy';
import { AbilityUnitBase } from './AbilityUnitBase';

export class AbilityUnitHero extends AbilityUnitBase {
    private mTable: UNIT_HERO | null = null;
    private mLevelTable: UNIT_HERO_LEVEL | null = null;
    private readonly mEquips: Map<EQUIP_SLOT_TYPE, AbilityItemEquip> = new Map();

    get unitId(): string { return this.mTable?.unit_id ?? ''; }
    get equips(): ReadonlyMap<EQUIP_SLOT_TYPE, AbilityItemEquip> { return this.mEquips; }

    init(table: UNIT_HERO, levelTable: UNIT_HERO_LEVEL): void {
        this.mTable = table;
        this.mLevelTable = levelTable;
        this.initUnitState(levelTable.unit_level, levelTable.max_hp);
    }

    equip(equip: AbilityItemEquip, slotType: EQUIP_SLOT_TYPE): boolean {
        if (!equip || slotType === EQUIP_SLOT_TYPE.NONE)
            return false;

        if (AbilityEquipSlotPolicy.getPlacementFailure(equip, slotType, this.mEquips) !== AbilityEquipPlacementFailure.None)
            return false;

        if (slotType === EQUIP_SLOT_TYPE.HAND_MAIN && AbilityEquipSlotPolicy.isTwoHanded(equip))
            this.unequip(EQUIP_SLOT_TYPE.HAND_SUB);

        const prev = this.mEquips.get(slotType);
        if (prev) {
            if (AbilityItemEquip.isSame(prev, equip)) {
                prev.setOwner(this.unitId, slotType);
                return true;
            }

            this.unequip(slotType);
        }

        const existingSlot = this.findEquipSlot(equip);
        if (existingSlot !== EQUIP_SLOT_TYPE.NONE) {
            if (existingSlot === slotType)
                return true;

            this.unequip(existingSlot);
        }

        if (equip.isEquipped)
            equip.clearOwner();

        this.mEquips.set(slotType, equip);
        equip.setOwner(this.unitId, slotType);
        this.applyEquipStats(equip, +1);
        return true;
    }

    unequip(slotType: EQUIP_SLOT_TYPE): boolean {
        const equip = this.mEquips.get(slotType);
        if (!equip)
            return false;

        this.applyEquipStats(equip, -1);

        if (equip.ownerUnitId === this.unitId && equip.ownerSlotType === slotType)
            equip.clearOwner();

        this.mEquips.delete(slotType);
        return true;
    }

    clone(): AbilityUnitHero {
        const c = new AbilityUnitHero();
        c.mTable = this.mTable;
        c.mLevelTable = this.mLevelTable;
        c.copyStatsFrom(this);
        c.copyUnitStateFrom(this);
        for (const [slot, equip] of this.mEquips)
            c.mEquips.set(slot, equip);
        return c;
    }

    private applyEquipStats(equip: AbilityItemEquip, sign: number): void {
        if (!equip || sign === 0)
            return;

        for (const [statType, statValue] of equip.stats) {
            if (!this.shouldApplyEquipStat(statType))
                continue;

            this.addStat(statType, statValue * sign);
        }
    }

    private findEquipSlot(equip: AbilityItemEquip): EQUIP_SLOT_TYPE {
        for (const [slotType, slottedEquip] of this.mEquips) {
            if (AbilityItemEquip.isSame(slottedEquip, equip))
                return slotType;
        }

        return EQUIP_SLOT_TYPE.NONE;
    }

    private shouldApplyEquipStat(statType: UNIT_STAT_TYPE): boolean {
        return statType !== UNIT_STAT_TYPE.NONE
            && statType !== UNIT_STAT_TYPE.ITEM_LEVEL
            && statType !== UNIT_STAT_TYPE.ITEM_AMOUNT;
    }
}

File: framework-cs/apps/UnityExample/Assets/Samples/Devian Foundation/0.1.0/GamePackage/Runtime/Ability/AbilityUnitHero.cs
using System.Collections.Generic;
using Devian.Domain.Game;

namespace Devian
{
    public sealed class AbilityUnitHero : AbilityUnitBase
    {
        UNIT_HERO mTable = null;
        UNIT_HERO_LEVEL mLevelTable = null;
        readonly Dictionary<EQUIP_SLOT_TYPE, AbilityItemEquip> mEquips = new();

        public override string UnitId => mTable?.unit_id ?? string.Empty;
        public IReadOnlyDictionary<EQUIP_SLOT_TYPE, AbilityItemEquip> Equips => mEquips;

        public void Init(UNIT_HERO table, UNIT_HERO_LEVEL levelTable)
        {
            mTable = table;
            mLevelTable = levelTable;

            if (levelTable == null)
                return;

            InitUnitState(levelTable.unit_level, levelTable.max_hp);
        }

        public override AbilityBase Clone()
        {
            var c = new AbilityUnitHero();
            c.mTable = mTable;
            c.mLevelTable = mLevelTable;
            c.CopyStatsFrom(this);
            c.CopyUnitStateFrom(this);
            foreach (var kv in mEquips)
                c.mEquips[kv.Key] = kv.Value;
            return c;
        }

        internal bool _Equip(AbilityItemEquip equip, EQUIP_SLOT_TYPE slotType)
        {
            if (equip == null || slotType == EQUIP_SLOT_TYPE.NONE)
                return false;

            if (AbilityEquipSlotPolicy.GetPlacementFailure(equip, slotType, mEquips) != AbilityEquipPlacementFailure.None)
                return false;

            if (slotType == EQUIP_SLOT_TYPE.HAND_MAIN && AbilityEquipSlotPolicy.IsTwoHanded(equip))
                _Unequip(EQUIP_SLOT_TYPE.HAND_SUB);

            if (mEquips.TryGetValue(slotType, out var prev))
            {
                if (AbilityItemEquip.IsSame(prev, equip))
                {
                    prev.SetOwner(UnitId, slotType);
                    return true;
                }

                _Unequip(slotType);
            }

            var existingSlot = findEquipSlot(equip);
            if (existingSlot != EQUIP_SLOT_TYPE.NONE)
            {
                if (existingSlot == slotType)
                    return true;

                _Unequip(existingSlot);
            }

            if (equip.IsEquipped)
                equip.ClearOwner();

            mEquips[slotType] = equip;
            equip.SetOwner(UnitId, slotType);
            applyEquipStats(equip, +1);
            return true;
        }

        internal bool _Unequip(EQUIP_SLOT_TYPE slotType)
        {
            if (!mEquips.TryGetValue(slotType, out var equip))
                return false;

            applyEquipStats(equip, -1);

            if (equip.OwnerUnitId == UnitId && equip.OwnerSlotType == slotType)
                equip.ClearOwner();

            mEquips.Remove(slotType);
            return true;
        }

        void applyEquipStats(AbilityItemEquip equip, int sign)
        {
            if (equip == null || sign == 0)
                return;

            foreach (var kv in equip.Stats)
            {
                if (!shouldApplyEquipStat(kv.Key))
                    continue;

                AddStat(kv.Key, kv.Value * sign);
            }
        }

        EQUIP_SLOT_TYPE findEquipSlot(AbilityItemEquip equip)
        {
            foreach (var kv in mEquips)
            {
                if (AbilityItemEquip.IsSame(kv.Value, equip))
                    return kv.Key;
            }

            return EQUIP_SLOT_TYPE.NONE;
        }

        static bool shouldApplyEquipStat(UNIT_STAT_TYPE statType)
        {
            return statType != UNIT_STAT_TYPE.NONE
                && statType != UNIT_STAT_TYPE.ITEM_LEVEL
                && statType != UNIT_STAT_TYPE.ITEM_AMOUNT;
        }
    }
}
